lib/std/os/linux/start_pie.zig

```diff
@@ -0,0 +1,138 @@
+const std = @import("std");
+const elf = std.elf;
+const builtin = @import("builtin");
+const assert = std.debug.assert;
+
+const R_AMD64_RELATIVE = 8;
+const R_386_RELATIVE = 8;
+const R_ARM_RELATIVE = 23;
+const R_AARCH64_RELATIVE = 1027;
+const R_RISCV_RELATIVE = 3;
+
+const ARCH_RELATIVE_RELOC = switch (builtin.arch) {
+    .i386 => R_386_RELATIVE,
+    .x86_64 => R_AMD64_RELATIVE,
+    .arm => R_ARM_RELATIVE,
+    .aarch64 => R_AARCH64_RELATIVE,
+    .riscv64 => R_RISCV_RELATIVE,
+    else => @compileError("unsupported architecture"),
+};
+
+// Just a convoluted (but necessary) way to obtain the address of the _DYNAMIC[]
+// vector as PC-relative so that we can use it before any relocation is applied
+fn getDynamicSymbol() [*]elf.Dyn {
+    const addr = switch (builtin.arch) {
+        .i386 => asm volatile (
+            \\ .weak _DYNAMIC
+            \\ .hidden _DYNAMIC
+            \\ call 1f
+            \\ 1: pop %[ret]
+            \\ lea _DYNAMIC-1b(%[ret]), %[ret]
+            : [ret] "=r" (-> usize)
+        ),
+        .x86_64 => asm volatile (
+            \\ .weak _DYNAMIC
+            \\ .hidden _DYNAMIC
+            \\ lea _DYNAMIC(%%rip), %[ret]
+            : [ret] "=r" (-> usize)
+        ),
+        // Work around the limited offset range of `ldr`
+        .arm => asm volatile (
+            \\ .weak _DYNAMIC
+            \\ .hidden _DYNAMIC
+            \\ ldr %[ret], 1f
+            \\ add %[ret], pc
+            \\ b 2f
+            \\ 1: .word _DYNAMIC-1b
+            \\ 2:
+            : [ret] "=r" (-> usize)
+        ),
+        // A simple `adr` is not enough as it has a limited offset range
+        .aarch64 => asm volatile (
+            \\ .weak _DYNAMIC
+            \\ .hidden _DYNAMIC
+            \\ adrp %[ret], _DYNAMIC
+            \\ add %[ret], %[ret], #:lo12:_DYNAMIC
+            : [ret] "=r" (-> usize)
+        ),
+        .riscv64 => asm volatile (
+            \\ lla %[ret], _DYNAMIC
+            : [ret] "=r" (-> usize)
+        ),
+        else => @compileError("???"),
+    };
+    if (addr == 0) unreachable;
+    return @intToPtr([*]elf.Dyn, addr);
+}
+
+pub fn apply_relocations() void {
```

Comment:
Case-styling? Or is this a traditional name from somewhere?

Reply:
I'm still not used to camelCase all the identifiers, will change it later :)